aten/src/ATen/mps/MPSStream.mm

```diff
@@ -185,40 +193,73 @@
 //  MPSEvent
 //-----------------------------------------------------------------
 
-MPSEvent::MPSEvent() {
-  _event = [MPSDevice::getInstance()->device() newSharedEvent];
+MPSEvent::MPSEvent(bool deferInitialization) :
+    is_initialized(false), _signalCounter(0), _stream(nil), _event(nil), _listener(nil) {
+  if (!deferInitialization) {
+    initialize();
+  }
 }
 
 MPSEvent::~MPSEvent() {
-  [_event release];
-  _event = nil;
-}
-
-void MPSEvent::recordEvent(MPSStream* stream) {
-  @autoreleasepool {
-    _isRecorded = true;
-    dispatch_sync(stream->queue(), ^() {
-      @autoreleasepool {
-        id<MTLCommandBuffer> commandBuffer = stream->commandBuffer();
-        [commandBuffer encodeSignalEvent:_event value:_currentValue];
-        stream->commit(true);
-      }
-    });
+  if (_event) {
+    [_event release];
+    _event = nil;
   }
+  if (_listener) {
+    [_listener release];
+    _listener = nil;
+  }
+}
+
+void MPSEvent::initialize() {
+  _stream = getDefaultMPSStream();
+  _event = [_stream->device() newSharedEvent];
+  _listener = [[MTLSharedEventListener alloc] init];
+  is_initialized = true;
 }
 
-void MPSEvent::waitForEvent(MPSStream* stream) {
-  dispatch_sync(stream->queue(), ^() {
+void MPSEvent::recordEvent(bool syncEvent) {
+  if (!is_initialized)
+    initialize();
+
+  dispatch_sync(_stream->queue(), ^() {
+    @autoreleasepool {
+      ++_signalCounter;
+      id<MTLCommandBuffer> commandBuffer = _stream->commandBuffer();
+      [commandBuffer encodeSignalEvent:_event value:_signalCounter];
+      if (syncEvent)
+        _stream->synchronize(SyncType::COMMIT);
+    }
+  });
+}
+
+void MPSEvent::waitForEvent(bool syncEvent) {
+  TORCH_INTERNAL_ASSERT(is_initialized);
+  dispatch_sync(_stream->queue(), ^() {
+    @autoreleasepool {
+      id<MTLCommandBuffer> commandBuffer = _stream->commandBuffer();
+      [commandBuffer encodeWaitForEvent:_event value:_signalCounter];
+      if (syncEvent)
+        _stream->synchronize(SyncType::COMMIT);
```

Comment:
I am curious how does this work with commit and continue. Since we don't know the Command buffer boundaries. Its going to be not the command buffer we expect.

Reply:
We shouldn't need to sync in MPSEvents. We'll pass false to syncEvent, in case we use MPSEvent in future. I just kept it here because it was there since the beginning and I made it optional via syncEvent boolean.